File: src/Sorting/lc0215_KthLargestElementInAnArray.exs
# https://leetcode.cn/problems/kth-largest-element-in-an-array/
defmodule Solution do
  @spec find_kth_largest(nums :: [integer], k :: integer) :: integer
  def find_kth_largest(nums, k) do
    true = is_integer(k) and 0 < k and k <= length(nums)

    tree =
      Enum.reduce(nums, :gb_trees.empty(), fn
        num, tree ->
          cond do
            :gb_trees.size(tree) == k ->
              {_, smallest_num} = :gb_trees.smallest(tree)

              cond do
                smallest_num < num ->
                  {_, _, temp_tree} = :gb_trees.take_smallest(tree)
                  :gb_trees.insert({num, :erlang.unique_integer()}, num, temp_tree)

                true ->
                  tree
              end

            true ->
              :gb_trees.insert({num, :erlang.unique_integer()}, num, tree)
          end
      end)

    {_, result} = :gb_trees.smallest(tree)
    result
  end
end

Solution.find_kth_largest([3, 2, 1, 5, 6, 4], 2)
|> dbg()

# http://erlang.org/pipermail/erlang-questions/2007-July/027720.html

# Subject: Re: [erlang-questions] Priority queue

# You can use a gb_trees structure:

# in(Item, Prio, Q) ->
#    gb_trees:insert({Prio,now()}, Item, Q).

# out(Q) ->
#    gb_trees:take_smallest(Q).

# peek(Q) ->
#    gb_trees:smallest(Q).

# The order will be FIFO. If you want LIFO, you
# could negate the Prio value and take_largest
# instead.

# https://learnyousomeerlang.com/time
# The new components of the VM are exposed to the user with the following functions:
# erlang:unique_integer() and erlang:unique_integer(Options), which returns unique values.

File: src/Sorting/lc0215_KthLargestElementInAnArray.py
# https://leetcode.cn/problems/kth-largest-element-in-an-array/
class Solution:
    def findKthLargest(self, nums: list[int], k: int) -> int:
        # # 应对 leetcode极端测试用例: 升序超大数组
        # import random
        # random.shuffle(nums)

        if k > len(nums):
            raise "Invalid Input"

        # 如果数组长度为10， 如果想要选择第3大的元素，那么数组从小大排序后这个元素所在的位置是索引7
        target_value = quick_select(nums, 0, len(nums) - 1, len(nums) - k)
        return target_value


# left_index（包含），right_index（包含）
def quick_select(nums: list[int], left_index: int, right_index: int, target_index: int) -> int:
    if left_index == right_index:
        return nums[left_index]

    left_pointer, right_pointer = left_index, right_index

    # 要求左指针掠过的值小于标兵，所以标兵所在的初始位置是个坑，而且是待填的第一个坑，
    # 所以先移动右指针。
    # 最后把标兵的值填充到中间的坑里。
    # 同一时刻仅有一个坑位。
    # 遇到坏元素停下来挖出来填到对方的坑里，然后换对方指针移动来填我的坑。
    to_compare_with = nums[left_pointer]
    while True:
        while left_pointer < right_pointer and to_compare_with <= nums[right_pointer]:
            right_pointer -= 1
        if left_pointer == right_pointer:
            break
        nums[left_pointer] = nums[right_pointer]
        left_pointer += 1
        while left_pointer < right_pointer and nums[left_pointer] < to_compare_with:
            left_pointer += 1
        if left_pointer == right_pointer:
            break
        nums[right_pointer] = nums[left_pointer]
        right_pointer -= 1
    assert left_pointer == right_pointer
    nums[left_pointer] = to_compare_with

    if target_index <= left_pointer:
        return quick_select(nums, left_index, left_pointer, target_index)
    else:
        return quick_select(nums, left_pointer + 1, right_index, target_index)


# 时间复杂度：O(n)，证明过程可以参考「《算法导论》9.2：期望为线性的选择算法」。
# 空间复杂度：O(logn)，递归使用栈空间的空间代价的期望为 O(logn)。